### PR TITLE
feat/chat notification

### PR DESCRIPTION
**Pull request by sieuhuflit**
## Overview

- The gateway is just in the middle to handle send push notifications
- Client implement registration token using this [Pusher set](https://spec.matrix.org/latest/client-server-api/#post_matrixclientv3pushersset) (similar to addFcmToken to our backend, but this for matrix side)
- When homeserver receives an event
- Homeserver forwards information about received events to the push gateway using this [POST /_matrix/push/v1/notify](https://matrix.org/docs/spec/push_gateway/latest#post-matrix-push-v1-notify)
- The gateway then submits a push notification to the push notification provider (e.g. APNS, GCM) using some configure
https://spec.matrix.org/unstable/push-gateway-api/
https://github.com/matrix-org/sygnal/blob/main/docs/applications.md#notification-flow

## Video explanation 
- https://www.loom.com/share/e743d92c69654df0ab6e7da6086bc8dc?sid=21bdb45d-e09e-47ff-a84c-24590b08f149

## Demo

Aug-10-2023 16-55-24

### CODE REVIEW

File: lib/core/presentation/pages/home/home_page.dart [hunk 1]
```diff
@@ -64,10 +67,22 @@ class _HomeListingView extends StatefulWidget {
...
+  final _appRouter = AppRouter();
+
   @override
   void initState() {
     super.initState();
     getIt<ShakeService>().startShakeDetection(context);
+
+    WidgetsBinding.instance.addPostFrameCallback((_) async {
+      getIt<MatrixService>().backgroundPush.setupContextAndRouter(
+            router: _appRouter,
```
Comment (mthinh) [inline lib/core/presentation/pages/home/home_page.dart:79]: please move this to _App widget.
Homescreen should not contain this logic

File: lib/core/service/firebase/firebase_service.dart [hunk 2]
```diff
@@ -87,11 +77,14 @@ class FirebaseService {
       initializationSettings,
       onDidReceiveNotificationResponse:
           (NotificationResponse notificationResponse) {
+        print('onDidReceiveNotificationResponse');
+        print(notificationResponse.payload.toString());
```
Comment (mthinh) [inline lib/core/service/firebase/firebase_service.dart:81]: should log in debug mode only

File: lib/core/utils/chat_notification/background_push.dart [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+import 'package:matrix/matrix.dart';
+
+import 'famedlysdk_store.dart';
+import 'push_helper.dart';
+
+// ignore: unused_element
+Future<void> _firebaseMessagingBackgroundHandler(RemoteMessage message) async {
+  if (kDebugMode) {
+    print('Handling a background message: ${message.messageId}');
+    print('Message data: ${message.data}');
```
Comment (RonTran636) [inline lib/core/utils/chat_notification/background_push.dart:24]: Prefer log debug over print

File: lib/core/utils/chat_notification/push_helper.dart [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+    platformChannelSpecifics,
+    payload: event.roomId,
+  );
+  Logs().v('Push helper has been completed!');
+}
+
+/// Workaround for the problem that local notification IDs must be int but we
+/// sort by [roomId] which is a String. To make sure that we don't have duplicated
+/// IDs we map the [roomId] to a number and store this number.
+Future<int> mapRoomIdToInt(String roomId) async {
```
Comment (RonTran636) [inline lib/core/utils/chat_notification/push_helper.dart:251]: If we need a roomId must be int, we can use `roomId.hashmap`. For each unique roomId will be a unique hashmap

**Review by RonTran636 — APPROVED** [review]
Need some minor changes but LGTM

File: lib/core/service/firebase/firebase_service.dart (hunk 2 above)
Comment (sieuhuflit) [reply lib/core/service/firebase/firebase_service.dart]: Moved 👍

File: lib/core/utils/chat_notification/background_push.dart (hunk 3 above)
Comment (sieuhuflit) [reply lib/core/utils/chat_notification/background_push.dart:24]: Removed

File: lib/core/presentation/pages/home/home_page.dart (hunk 1 above)
Comment (sieuhuflit) [reply lib/core/presentation/pages/home/home_page.dart]: Resolved

File: lib/core/utils/chat_notification/push_helper.dart (hunk 4 above)
Comment (sieuhuflit) [reply lib/core/utils/chat_notification/push_helper.dart:251]: I see, good idea for improvement, this code I keep it from FluffyChat, not self-write this code, and purpose I keep this for easier to improve or monitor when FluffyChat have some stuff changed.

**Review by mthinh — APPROVED** [review]

**Pull request merged**